storage/remote/sigv4.go

```diff
@@ -0,0 +1,138 @@
+// Copyright 2021 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package remote
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/textproto"
+	"sync"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/credentials/stscreds"
+	"github.com/aws/aws-sdk-go/aws/session"
+	signer "github.com/aws/aws-sdk-go/aws/signer/v4"
+	"github.com/prometheus/prometheus/config"
+)
+
+var sigv4HeaderDenylist = []string{
+	"uber-trace-id",
+}
+
+type sigV4RoundTripper struct {
+	region string
+	next   http.RoundTripper
+	pool   sync.Pool
+
+	signer *signer.Signer
+}
+
+// newSigV4RoundTripper returns a new http.RoundTripper that will sign requests
+// using Amazon's Signature Verification V4 signing procedure. The request will
+// then be handed off to the next RoundTripper provided by next. If next is nil,
+// http.DefaultTransport will be used.
+//
+// Credentials for signing are retrieved using the the default AWS credential
+// chain. If credentials cannot be found, an error will be returned.
+func newSigV4RoundTripper(cfg *config.SigV4Config, next http.RoundTripper) (http.RoundTripper, error) {
+	if next == nil {
+		next = http.DefaultTransport
+	}
+
+	creds := credentials.NewStaticCredentials(cfg.AccessKey, string(cfg.SecretKey), "")
```

Comment:
Access and secret are not the preferred way of authenticating, instead we use the AWS SDK default credential provider chain by default.

```
sess, err := session.NewSession()
```

returns a session with the default credentials. sess.Config has credentials and the default region.

See https://github.com/aws-observability/amp-eks-iam/blob/main/main.go#L141 as an example of how to read the default region.

AWS users never hardcode credentials and it's considered an anti-pattern. We shouldn't ask them to create access/secret keys.

Reply:
I agree. It's not a hard requirement here, and if you don't provide the keys, it'll fall back to using the credentials chain (see below on line 59-69). 

My understanding was that Prometheus team wants to still support optionally providing credentials to be consistent with the EC2 SD, which functions the same way with the fallback to the default chain when left blank.

Reply:
Ah, gotcha. Allowing the credentials in addition to the default credentials sounds great.

Reply:
Yes, as we support this pattern in the EC2 service discovery, I prefer to let those user "easily" use to SIGv4, with the same pattern they use for EC2 service discovery.

Reply:
By the way, just wonder if the security token (3rd argument to `NewStaticCredentials`) is intentionally left blank to not support temporary credentials when customer config credentials through `prometheus.yaml`?